### PR TITLE
[Samuel Huang] Duke Increments

### PR DESCRIPTION
**Pull request by Eclmist**

### CODE REVIEW

File: src/main/java/duke/Duke.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    command.execute(tasks, ui);
+                    shouldExit = command.isExitCommand();
+                }
+
+            } catch (DukeException e) {
+                ui.printError(e.getMessage());
+            }
+        }
+    }
+}
```
Comment (souwmyaa) [inline src/main/java/duke/Duke.java:31]: Code is well written!

File: src/main/java/duke/command/AddCommand.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+import duke.exception.DukeException;
+import duke.io.Ui;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.task.Todo;
+
+public class AddCommand extends Command {
```
Comment (souwmyaa) [inline src/main/java/duke/command/AddCommand.java:10]: Using inheritance here is a good idea! An alternate suggestion would be to have an enum for commands.

File: src/main/java/duke/command/DeleteCommand.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public void execute(TaskList taskList, Ui ui) throws DukeException {
+        DukeException.throwIf(!taskList.isIndexValid(indexToDelete), "The input index is out of bounds!");
+        Task deletedTask = taskList.removeAtIndex(indexToDelete);
+        ui.stylizedPrint(
+                "You have removed the following task:\n",
+                "\t" + deletedTask.toString(),
+                String.format("You now have %d task(s) in the list.", taskList.size()));
+    }
+}
```
Comment (souwmyaa) [inline src/main/java/duke/command/DeleteCommand.java:33]: Code is well written!

File: src/main/java/duke/command/ExitCommand.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    public void execute(TaskList taskList, Ui ui) {
+        Serializer.serialize(taskList);
+        ui.printGoodbyeMessage();
+    }
+
+    @Override
+    public boolean isExitCommand() {
+        return true;
+    }
+}
```
Comment (souwmyaa) [inline src/main/java/duke/command/ExitCommand.java:24]: Code is well written!

File: src/main/java/duke/task/Task.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    public String toString() {
+        return getTypeIcon() + getStatusIcon() + " " + description;
+    }
+
+    public enum TaskType {
+        TASK_TYPE_TODO,
+        TASK_TYPE_EVENT,
+        TASK_TYPE_DEADLINE
+    }
+}
```
Comment (souwmyaa) [inline src/main/java/duke/task/Task.java:109]: Code is well written!

**Review by souwmyaa — COMMENTED** [review]
Your code is extremely organised and well written!

File: src/main/java/duke/io/Ui.java [hunk 6]
```diff
@@ -37,14 +58,28 @@ public void printError(String... error) {
...
     public void printHorizontalBar() {
         System.out.println(OUTPUT_HORIZONTAL_LINE);
     }
 
+    /**
+     * Prints one or more lines of text with ASCII accents.
+     *
+     * @param lines A variable number of strings, each would print a line of text.
+     */
     public void stylizedPrint(String... lines) {
```
Comment (Hardoh) [inline src/main/java/duke/io/Ui.java]: Hi Sam! As both stylizedPrint methods "print" stuff, would it be better to name them with verbs (similar to your printHorizontalBar), i.e. printStylized? Cheers!

File: src/main/java/duke/task/TaskList.java [hunk 7]
```diff
@@ -67,7 +67,7 @@ public boolean isIndexValid(String index) {
      * element in the taskList. This function returns all the names of tasks (from their respective
      * toString() functions) as an array.
      *
-     * @return An array containing the names of every task in the taskList.
+     * @return An array containing the names of every duke.task in the taskList.
```
Comment (Hardoh) [inline src/main/java/duke/task/TaskList.java]: Hi Sam, disregard this if this was your intention. But, perhaps reflecting the package of task in this comment might be unnecessary and possible a result of refactoring when you packaged your classes. Consider inspecting and removing the "duke." in "duke.task", perhaps?

File: src/main/java/duke/io/Ui.java (hunk 6 above)
Comment (Eclmist) [reply src/main/java/duke/io/Ui.java]: Hi Hardy :)
`printStylized()` does sound better, good suggestion

File: src/main/java/duke/task/TaskList.java (hunk 7 above)
Comment (Eclmist) [reply src/main/java/duke/task/TaskList.java]: This was indeed the result of a rogue refactor, nice catch!

File: src/main/java/duke/Duke.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
+package duke;
+
+import duke.command.Command;
+import duke.exception.DukeException;
+import duke.io.Parser;
+import duke.io.Serializer;
+import duke.io.Ui;
+import duke.task.TaskList;
+
```
Comment (foochifa) [inline src/main/java/duke/Duke.java:9]: Nicely written with good breaks to enhance readability! Just missing javadocs!

File: src/main/java/duke/io/Parser.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            return new AddCommand(Task.TaskType.TASK_TYPE_DEADLINE, args);
+        case "event":
+            return new AddCommand(Task.TaskType.TASK_TYPE_EVENT, args);
+        case "delete":
+            return new DeleteCommand(operationIndex);
+        case "done":
+            return new DoneCommand(operationIndex);
+        default:
+            throw new DukeException("'" + command + "' is not a recognized command." +
+                    "\n\tYou are advised to stop trying to break the system.");
```
Comment (foochifa) [inline src/main/java/duke/io/Parser.java]: I think you may have missed out all the breaks in your cases!

File: src/main/java/duke/task/Task.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package duke.task;
+
+import java.io.Serializable;
```
Comment (foochifa) [inline src/main/java/duke/task/Task.java:3]: Neat! Great idea in using serializable.

File: src/main/java/duke/task/Event.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
+package duke.task;
+
+public class Event extends Task {
+
+    // TODO: Implement start/end time
+    protected String time;
```
Comment (foochifa) [inline src/main/java/duke/task/Event.java:6]: Maybe it will be good to use LocalTime?

File: src/main/java/duke/io/Ui.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
+package duke.io;
+
+import java.io.InputStream;
+import java.util.Scanner;
+
+public class Ui {
```
Comment (foochifa) [inline src/main/java/duke/io/Ui.java]: Really good use of the Ui class!

File: src/main/java/duke/io/Parser.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        String command = inputs[0];
+        String args = inputs.length > 1 ? inputs[1] : "";
+        int operationIndex;
+        try {
+            operationIndex = Integer.parseInt(args) - 1; // Human index starts from 1. Converting to real index.
+        } catch (NumberFormatException e) {
+            operationIndex = -1;
+        }
+
+        switch (command) {
```
Comment (foochifa) [inline src/main/java/duke/io/Parser.java:34]: Maybe you could use an Enum class here instead of just String!

**Review by foochifa — COMMENTED** [review]
Very well written code! Clean with good breaks and very little code style violations. The classes were also nicely packaged! Also, really interesting to see how you used serializable!

File: src/main/java/duke/io/Parser.java (hunk 9 above)
Comment (foochifa) [reply src/main/java/duke/io/Parser.java]: Sorry I just realised they are all return statements! Please ignore my above comment!

**Comment by Eclmist** [comment]
@foochifa @souwmyaa @Hardoh Thanks for the review :)

**Comment by j-lum** [comment]
Interesting implementation of fuzzy search.

I know we said you can add some personality to Duke but leave the vulgarities to me.

**Pull request closed without merging**